### DISCUSSION TITLE
ESP32-S2 wifi causing neopixel flickering

### DISCUSSION THREAD

**Discussion by DougWilkinson** [micropython/micropython#15565] [ESP32]
Hi,
I picked up some esp32-S2FN4R2 4MB Flash 2MB PSRAM boards (generic mini) recently and have noticed  flickering when wifi is enabled and neopixels are being updated. **If I don't enable wifi, it works fine.** 

The same code runs fine on an esp8266 with wifi enabled.

I'm running v1.23.0 downloaded here:
https://micropython.org/resources/firmware/ESP32_GENERIC_S2-20240602-v1.23.0.bin

I've tried v1.20.0 and v1.19.1 with the same results.

The code that I'm running has several underlying modules and is using async, but the one that updates neopixels is:

https://github.com/DougWilkinson/ams/blob/master/ledclock.py

Note: I added a Timer to perform the led.write() to see if updating while inside an ISR would help. It didn't and was not my original approach to this in case you are wondering.

I've searched everywhere for things to try (like increasing RMT RAM?) but not sure if that can be done or is applicable.

As I understand it, micropython runs as an application on RTOS that handles things like wifi, so is the S2 just a bad chip to use in this case? Any other ideas?

Thank you,
--Doug

**Comment by davefes** [comment]
Wifi is enabled ... are you connected to some other device and there is traffic?  If so, I would put an oscilloscope on 3V3 and see if there are voltage dips.

**Reply by DougWilkinson** [reply]
Yes, I'm using MQTT to receive time updates. I tried connecting the 5v for the LEDS to a separate power source and still see the same issue. It is a string of 116 LEDs, but no more than 20 of them are on at any given time. (and thank you @scruss for your suggestion as well) -- Are you running the S2 version of the esp32?

**Reply by Josverl** [reply]
Assuming you are using neopixels, where each have 3 or 4 actual leds drawing power, if seen issues with strings over 15 pixels leading to brownouts on the board.

I'd definitely recommend powering the strip independently, connect the ground though.

If you don't want that, then it's time to use brains + some software to reduce the power consumption, 

A good read is: https://learn.adafruit.com/sipping-power-with-neopixels/insights

**Reply by DougWilkinson** [reply]
@Josverl yes, I did power separately (with grounds connected) and still see the same flickering. I am thinking it is not a power issue at this point (although I understand that bad power can cause a lot of strange issues). Any other ideas? I'm still wondering if the neopixel write() is being interrupted with wifi being the primary suspect.

**Reply by scruss** [reply]
Do the LEDs flicker when the wifi network is accessed and the LEDs are displaying a static display?

I just tried with a Wmoes S2 Mini, the Wemos tiny 64 Neopixel shield and MicroPython 1.23. I set all 64 pixels on, then ran [µPing](https://gist.github.com/shawwwn/91cc8979e33e82af6d99ec34c38195fb).ping() my local router once per second. There was no flickering I could see in the pixels

**Reply by Josverl** [reply]
I'm on vacation and don't have the hardware to repro, but I was wondering if the neopixel's data wire might pick up interference from the wifi if it's too close or looped.
I have a few pieces of shielded wire in the tinkerbox to rule that out.

**Comment by scruss** [comment]
How are you powering the LEDs? You might need an additional PSU for them if wifi is drawing too much from USB.

**Comment by bixb922** [comment]
If "flicker" means irregular behaviour, then this could be due to WiFi operations blocking the process for some time. Although there is a second CPU and RTOS, I see on the ESP32-S3 that the wifi network operations even with asyncio block the primary CPU.
